File: backend/src/services/transactionService.js
const { TransactionResponse } = require('../models/Transaction');
const supabase = require('../config/supabaseClient');

/**
 * Generate an ISO8583 message based on transaction data
 * @param {Object} transaction - Transaction data
 * @param {String} responseCode - Response code
 * @returns {Object} ISO8583 message
 */
function generateIso8583Message(transaction, responseCode) {
  const now = new Date();
  
  return {
    mti: "0110", // Message Type Indicator
    primaryAccountNumber: transaction.card_number,
    processingCode: "000000", // Example processing code
    amount: transaction.amount,
    transmissionDateTime: now.toISOString().replace(/[-:T.Z]/g, '').substring(0, 14),
    systemTraceNumber: Math.floor(100000 + Math.random() * 900000).toString(), // Random 6-digit number
    localTransactionTime: now.toLocaleTimeString('en-US'),
    localTransactionDate: now.toLocaleDateString('en-US'),
    merchantType: "5999", // Example merchant type
    responseCode: responseCode,
    terminalId: "TERM001", // Example terminal ID
    merchantId: transaction.merchant_id
  };
}

/**
 * Process a transaction
 * @param {Object} transaction - Transaction request data
 * @returns {Promise<TransactionResponse>} Transaction response
 */
async function processTransaction(transaction) {
  // Simulate processing delay
  await new Promise(resolve => setTimeout(resolve, Math.random() * 900 + 100)); // 100-1000ms

  // Simple validation and approval logic
  const isApproved = Math.random() < 0.9; // 90% success rate
  const responseCode = isApproved ? "00" : "05";
  const authorizationCode = isApproved 
    ? Array.from({ length: 6 }, () => "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"[Math.floor(Math.random() * 36)]).join('')
    : null;

  // Generate ISO 8583 message
  const iso8583Message = generateIso8583Message(transaction, responseCode);

  const now = new Date().toISOString();
  
  // Prepare transaction data for Supabase
  const transactionData = {
    card_number: transaction.card_number,
    amount: transaction.amount,
    merchant_id: transaction.merchant_id,
    status: isApproved ? "APPROVED" : "DECLINED",
    type: "PURCHASE",
    timestamp: now,
    created_at: now,
    updated_at: now,
    iso8583_message: iso8583Message
  };

  // Store transaction in Supabase
  try {
    await supabase.from("transactions").insert(transactionData);
  } catch (error) {
    console.error("Error storing transaction:", error);
    // Continue even if storage fails
  }

  return new TransactionResponse({
    success: isApproved,
    message: isApproved ? "Transaction approved" : "Transaction declined",
    data: transaction,
    response_code: responseCode,
    authorization_code: authorizationCode
  });
}

module.exports = {
  processTransaction,
  generateIso8583Message
};

File: backend-old/app/services/transaction_service.py
from datetime import datetime
from app.models.transaction import TransactionResponse
from app.utils.supabase_client import supabase
import random
import time

def generate_iso8583_message(transaction, response_code):
    """
    Generate an ISO 8583 message based on the transaction data.
    
    :param transaction: The transaction data (dict or Pydantic model)
    :param response_code: The response code (e.g., "00" for approved, "05" for declined)
    :return: ISO 8583 message as a dictionary
    """
    return {
        "mti": "0110",  # Message Type Indicator (0110 for financial transaction response)
        "primaryAccountNumber": transaction.card_number,
        "processingCode": "000000",  # Example processing code
        "amount": transaction.amount,
        "transmissionDateTime": datetime.now().strftime("%Y%m%d%H%M%S"),  # Current timestamp
        "systemTraceNumber": str(random.randint(100000, 999999)),  # Random trace number
        "localTransactionTime": datetime.now().strftime("%H:%M:%S"),  # Current time
        "localTransactionDate": datetime.now().strftime("%m/%d/%Y"),  # Current date
        "merchantType": "5999",  # Example merchant type
        "responseCode": response_code,
        "terminalId": "TERM001",  # Example terminal ID
        "merchantId": transaction.merchant_id,
    }

def process_transaction(transaction) -> TransactionResponse:
    # Simulate processing delay
    time.sleep(random.uniform(0.1, 1.0))

    # Simple validation and approval logic
    is_approved = random.random() < 0.9  # 90% success rate
    response_code = "00" if is_approved else "05"
    authorization_code = (
        "".join(random.choices("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789", k=6))
        if is_approved
        else None
    )

    # Generate ISO 8583 message
    iso8583_message = generate_iso8583_message(transaction, response_code)

    # Prepare transaction data for Supabase
    transaction_data = {
        "card_number": transaction.card_number,
        "amount": transaction.amount,
        "merchant_id": transaction.merchant_id,
        "status": "APPROVED" if is_approved else "DECLINED",
        "type": "PURCHASE",
        "timestamp": datetime.now().isoformat(),
        "created_at": datetime.now().isoformat(),
        "updated_at": datetime.now().isoformat(),
        "iso8583_message": iso8583_message,
    }

    # Store transaction in Supabase
    supabase.table("transactions").insert(transaction_data).execute()

    return TransactionResponse(
        success=is_approved,
        message="Transaction approved" if is_approved else "Transaction declined",
        data=transaction.dict(),
        response_code=response_code,
        authorization_code=authorization_code,
    )